GitUI/BranchTreePanel/RepoObjectsTree.Nodes.Submodules.cs

```diff
@@ -0,0 +1,442 @@
+                }
+
+                TreeViewNode.Text = "Submodules";
+
+                ThreadHelper.JoinableTaskFactory.RunAsync(async () => await LoadNodeDetailsAsync(CurrentReloadCancellationToken)).FileAndForget();
+            }
+
+            private void FillSubmoduleTree(SubmoduleInfoResult result)
+            {
+                _topProjectInfo = result.TopProject;
+
+                var nodes = new List<SubmoduleNode>();
+
+                // We always want to display submodules rooted from the top project. If the currently open project is the top project,
+                // OurSubmodules contains all child submodules recursively; otherwise, if we're currently in a submodule, SuperSubmodules
+                // contains all submodule info relative to the top project.
+                if (result.SuperSubmodules?.Count > 0)
+                {
+                    CreateSubmoduleNodes(result.SuperSubmodules, ref nodes);
+                }
+                else
+                {
+                    CreateSubmoduleNodes(result.OurSubmodules, ref nodes);
+                }
+
+                AddNodesToTree(nodes);
+            }
+
+            private void CreateSubmoduleNodes(IEnumerable<SubmoduleInfo> submodules, ref List<SubmoduleNode> nodes)
+            {
+                // result.OurSubmodules/SuperSubmodules contain a recursive list of submodules, but don't provide info about the super
+                // project path. So we deduce these by substring matching paths against an ordered list of all paths.
+                var modulePaths = submodules.Select(info => info.Path).ToList();
+
+                // Add current and parent module paths
+                var parentModule = Module;
+                while (parentModule != null)
+                {
+                    modulePaths.Add(parentModule.WorkingDir);
+                    parentModule = parentModule.SuperprojectModule;
+                }
+
+                // Sort descending so we find the nearest outer folder first
+                modulePaths = modulePaths.OrderByDescending(path => path).ToList();
+
+                string GetSubmoduleSuperPath(string submodulePath)
+                {
+                    var superPath = modulePaths.Find(path => submodulePath != path && submodulePath.Contains(path));
+                    Trace.Assert(superPath != null);
+                    return superPath;
+                }
+
+                foreach (var submoduleInfo in submodules)
+                {
+                    string superPath = GetSubmoduleSuperPath(submoduleInfo.Path);
+                    string localPath = PathUtil.GetDirectoryName(submoduleInfo.Path.Substring(superPath.Length).ToPosixPath());
+
+                    var isCurrent = submoduleInfo.Bold;
+                    nodes.Add(new SubmoduleNode(this, submoduleInfo, isCurrent, localPath, superPath));
+                }
+            }
+
+            private GitModule GetTopModule()
```

Comment:
```suggestion
            private IGitModule GetTopModule()
```

Reply:
Changing GitModule to IGitModule implies a bunch of other changes. For instance, SuperprojectModule is not a member of IGitModule. Should I go through and refactor all of this now?